p2p/ipld/retrieve.go

```diff
@@ -0,0 +1,235 @@
+// TODO @renaynay: move this to another package
+package ipld
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/ipfs/go-cid"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/nmt"
+	"github.com/lazyledger/nmt/namespace"
+)
+
+var (
+	ErrNotFoundInRange = fmt.Errorf("namespaceID not found in range")
+	ErrBelowRange      = fmt.Errorf("namespaceID is below minimum namespace ID in range")
+	ErrExceedsRange    = fmt.Errorf("namespaceID exceeds maximum namespace ID in range")
+)
+
+// RetrieveShares retrieves share data for the given namespace ID using the given DataAvailabilityHeader.
+func RetrieveShares(ctx context.Context, nID namespace.ID, dah *types.DataAvailabilityHeader, api coreiface.CoreAPI) ([][]byte, error) {
```

Comment:
We should use [DAGService](https://github.com/ipfs/go-ipld-format/blob/master/merkledag.go#L54) here as well now. It is a lower-level abstraction, and using it is generally more efficient and additionally, it enables optimized Bitswap code-paths through [Sessions](https://github.com/ipfs/go-merkledag/blob/master/session.go#L16). 

Unfortunately, we don't use CoreAPI in master anymore, so this has to be changed. If that does not clear how to reimplement logic with DAGService, ping me up and I will help with an explanation/implementation on any preferred language 😉